### DISCUSSION TITLE
winget install/upgrade output codes not documented

### DISCUSSION THREAD

**Discussion by ITAsADayJob1** [microsoft/winget-cli#3197] [General] [answered]
I'm using PowerShell to install apps via winget, and want to use the exit codes to understand the status, especially on failure.  Not the exit codes of the applications, but winget itself.

In my script I'm using:
`winget install --id $AppID --source msstore -e --silent --accept-package-agreements --accept-source-agreements`

Then using $LastExitCode to get the output

The exit codes are not standard, and are not documented.  What I have found so far are:
0: Successfully installed
-1978335157: "Failed when opening source(s); try the 'source reset' command if problem persists."
-1978335212: "No package found matching input criteria." during install, or "No installed package found matching input criteria." during upgrade
-1978335189: "No Applicable Upgrade Found."

Getting a lot of "-1978335205", on devices at various sites in one of our customers' Intune tenants, but I can't replicate to find out what this means.

What I got from ChatGPT as the expected exit codes:
0: The operation completed successfully.
1: An unknown error occurred.
2: The command line arguments were invalid.
3: The requested package could not be found.
4: The requested package is already installed.
5: The requested package could not be installed due to a dependency issue.
6: The requested package could not be uninstalled because it is not currently installed.
7: The requested package could not be updated because it is not currently installed.
8: The requested package update failed due to a dependency issue.
9: The requested package could not be updated because the package version is already the latest.
10: The requested package could not be updated because the package is locked by another process.

I just want a documented list of exit codes and what they mean so that we can understand what is going on with these failures.  Any chance of getting this?

**Comment by mdanish-kh** [comment, marked as answer]
You should put a 👍 on the related issue to raise its priority:
- https://github.com/microsoft/winget-cli/issues/1775

While they aren't documented, they are defined in this file [AppInstallerErrors.h](https://github.com/microsoft/winget-cli/blob/master/src/AppInstallerSharedLib/Public/AppInstallerErrors.h) and you can get a rough understanding of the error through its name.

Your specific exit code `-1978335205` corresponds to a hex value of `0x8A15001B` (signed 2's complement). Looking at the file, it seems an issue regarding Microsoft Store client being blocked by Group policy
https://github.com/microsoft/winget-cli/blob/84d3e87847edcd9faa3985f6b84a099d95c913bd/src/AppInstallerSharedLib/Public/AppInstallerErrors.h#L42

**Reply by ITAsADayJob1** [reply]
Thanks.  Exactly what I wanted.

**Comment by mdanish-kh** [comment]
Putting this here for any future explorers:

- Table of error codes was added in PR https://github.com/microsoft/winget-cli/pull/3202

https://github.com/microsoft/winget-cli/blob/master/doc/windows/package-manager/winget/returnCodes.md
